lib/combine_popolo_memberships.rb

```diff
@@ -1,30 +1,50 @@
 require 'combine_popolo_memberships/version'
+require 'date-range'
 
 # Takes multiple popolo membership arrays and combines them based on date.
 module CombinePopoloMemberships
-  def self.overlap(mem, term)
-    mS = mem[:start_date].to_s.empty? ? '0000-00-00' : mem[:start_date].to_s
-    mE = mem[:end_date].to_s.empty? ? '9999-99-99' : mem[:end_date].to_s
-    tS = term[:start_date].to_s.empty? ? '0000-00-00' : term[:start_date].to_s
-    tE = term[:end_date].to_s.empty? ? '9999-99-99' : term[:end_date].to_s
-
-    return unless mS < tE && mE > tS
-    (s, e) = [mS, mE, tS, tE].sort[1, 2]
-    {
-      _data: [mem, term],
-      start_date: s == '0000-00-00' ? nil : s,
-      end_date:   e == '9999-99-99' ? nil : e
-    }
+  class Membership
+    def initialize(raw)
+      @raw = raw
+    end
+
+    def start_date
+      @start_date ||= raw[:start_date].to_s.empty? ? '0000-00-00' : raw[:start_date].to_s
+    end
+
+    def end_date
+      @end_date ||= raw[:end_date].to_s.empty? ? '9999-99-99' : raw[:end_date].to_s
+    end
+
+    def date_range
+      @date_range ||= DateRange.new(start_date, end_date)
+    end
+
+    def overlap(other)
+      o = date_range.overlap(other.date_range)
+      return unless o
+      {}.merge(to_h).merge(other.to_h).tap do |h|
+        h[:start_date] = o.start_date
+        h[:end_date] = o.end_date
+      end
+    end
+
+    def to_h
+      raw
+    end
+
+    private
+
+    attr_reader :raw
   end
 
   def self.combine(h)
     into_name, into_data, from_name, from_data = h.flatten
-    from_data.product(into_data).map { |a, b| overlap(a, b) }.compact.map do |h|
-      from, to = h.delete :_data
-      [from, to].each { |orig|
-        orig.each { |k, v| h[k] = v unless %i(id start_date end_date).include? k }
-      }
+    from_data.product(into_data).map do |from, to|
+      h = Membership.new(from).overlap(Membership.new(to))
+      next unless h
+      h.delete(:id)
       h.merge(from_name => from[:id], into_name => to[:id])
```

Comment:
I don't really follow your thinking here, I'm afraid. The `Membership` class is essentially a private implementation — it doesn't _have_ a public interface. The only reason it seems to be exposed at all is so that you can unit test it (which I'm not sure has a lot of value in its own right, and definitely doesn't if it's already making things this rigid.)

It's not clear to me that that the `Membership` class will ever be directly part of the public interface — it's certainly not obvious so far that it will, though perhaps you've already thought through, or even have WIP code for, the next step. (If that _is_ the plan, then it worries me slightly that `Membership` may not be the right abstraction here, as one of the common usage patterns here is for one of the data lists to be a list of terms, which aren't really memberships)